flang/lib/Lower/OpenMP/OpenMP.cpp

```diff
@@ -2889,9 +2889,55 @@ static void genAtomicRead(lower::AbstractConverter &converter,
       fir::getBase(converter.genExprAddr(fromExpr, stmtCtx));
   mlir::Value toAddress = fir::getBase(converter.genExprAddr(
       *semantics::GetExpr(assignmentStmtVariable), stmtCtx));
-  genAtomicCaptureStatement(converter, fromAddress, toAddress,
-                            leftHandClauseList, rightHandClauseList,
-                            elementType, loc);
+
+  if (fromAddress.getType() != toAddress.getType()) {
```

Comment:
Please add commenst for:
-> Why we cannot use the typedAssignment lowering and is using custom lowering here? 
-> Why do these casts have to be added?
-> Why is it safe to do so?
-> Why we cannot use the typedAssignment lowering?

Reply:
Thanks. I have added an explanation for these